x-pack/plugins/observability_solution/apm/public/components/app/transaction_details/waterfall_with_summary/waterfall_container/waterfall/waterfall_helpers/waterfall_helpers.ts

```diff
@@ -462,3 +478,178 @@ export function getWaterfall(apiResponse: TraceAPIResponse): IWaterfall {
+    showCriticalPath: boolean;
+  };
+}) {
+  const tree = { ...root };
+  const queue: IWaterfallNode[] = [tree];
+
+  for (let queueIndex = 0; queueIndex < queue.length; queueIndex++) {
+    const node = queue[queueIndex];
+
+    const children = getChildren({ path, waterfall, waterfallItemId: node.item.id });
+    // Set childrenToLoad for all nodes enqueued.
+    // this allows lazy loading of child nodes
+    node.childrenToLoad = children.length;
+
+    if (maxLevelOpen > node.level) {
+      children.forEach((child, index) => {
+        const level = node.level + 1;
+
+        const currentNode: IWaterfallNode = {
+          id: btoa(`${node.id}-${child.id}-${index}`),
+          item: child,
+          children: [],
+          level,
+          expanded: level < maxLevelOpen,
+          childrenToLoad: 0,
+        };
+
+        node.children.push(currentNode);
+        queue.push(currentNode);
+      });
+    }
+  }
+
+  return tree;
+}
+
+export function buildTraceTree({
+  waterfall,
+  maxLevelOpen,
+  isOpen,
+  path,
+}: {
+  waterfall: IWaterfall;
+  maxLevelOpen: number;
+  isOpen: boolean;
+  path: {
+    criticalPathSegmentsById: Dictionary<CriticalPathSegment[]>;
+    showCriticalPath: boolean;
+  };
+}): IWaterfallNode | null {
+  const entry = waterfall.entryWaterfallTransaction;
+  if (!entry) {
+    return null;
+  }
+
+  const root: IWaterfallNode = {
+    id: entry.id,
+    item: entry,
+    children: [],
+    level: 0,
+    expanded: isOpen,
+    childrenToLoad: 0,
+  };
+
+  return buildTree({ root, maxLevelOpen, waterfall, path });
+}
+
+export const convertTreeToList = (root: IWaterfallNode | null): IWaterfallNodeFlatten[] => {
+  if (!root) {
+    return [];
+  }
+
+  const result: IWaterfallNodeFlatten[] = [];
+  const stack: IWaterfallNode[] = [root];
+
+  while (stack.length > 0) {
+    const node = stack.pop()!;
+
+    const { children, ...nodeWithoutChildren } = node;
+    result.push(nodeWithoutChildren);
+
+    if (node.expanded) {
+      for (let i = node.children.length - 1; i >= 0; i--) {
+        stack.push(node.children[i]);
+      }
+    }
+  }
+
+  return result;
+};
+
+export const updateTraceTreeNode = ({
+  root,
+  updatedNode,
+  waterfall,
+  path,
+}: {
+  root: IWaterfallNode;
+  updatedNode: IWaterfallNodeFlatten;
+  waterfall: IWaterfall;
+  path: {
+    criticalPathSegmentsById: Dictionary<CriticalPathSegment[]>;
+    showCriticalPath: boolean;
+  };
+}) => {
+  if (!root) {
+    return;
+  }
+
+  let tree = { ...root };
+  const stack: Array<{ parent: IWaterfallNode | null; index: number; node: IWaterfallNode }> = [
+    { parent: null, index: 0, node: root },
+  ];
+
+  while (stack.length > 0) {
+    const { parent, index, node } = stack.pop()!;
+
+    if (node.id === updatedNode.id) {
+      const newNode = { ...node, ...updatedNode };
+
+      if (updatedNode.expanded && updatedNode.childrenToLoad !== node.children.length) {
```

Comment:
Is checking on length here enough? Are there other kinds of changes that we should take into account? I was wondering if this can be a boolean instead? Or are there cases where these are both non-zero, but not equal?

Reply:
what I mean is that it might be more clear to use something like: 

`if (updatedNode.expanded && !updatedNode.hasInitializedChildren) {`